synapse/lib/logging.py

```diff
@@ -0,0 +1,265 @@
+import os
+import sys
+import asyncio
+import logging
+import weakref
+import collections
+
+import synapse.exc as s_exc
+import synapse.common as s_common
+
+import synapse.lib.coro as s_coro
+import synapse.lib.json as s_json
+import synapse.lib.const as s_const
+import synapse.lib.scope as s_scope
+
+logger = logging.getLogger(__name__)
+
+logtodo = []
+logbase = None
+loglock = asyncio.Lock()
+logevnt = asyncio.Event()
+logwindows = weakref.WeakSet()
+
+logfifo = collections.deque(maxlen=1000)
+
+def _addLogInfo(info):
+    logfifo.append(info)
+    if logbase is not None:
+        logtodo.append(info)
+        logevnt.set()
+
+async def _feedLogTask():
+
+    while not logbase.isfini:
+
+        await logevnt.wait()
+
+        if logbase.isfini:
+            return
+
+        todo = list(logtodo)
+
+        logevnt.clear()
+        logtodo.clear()
+
+        for wind in logwindows:
+            await wind.puts(todo)
+
+_glob_loginfo = {}
+def setLogInfo(name, valu):
+    '''
+    Configure global values which should be added to every log.
+    '''
+    _glob_loginfo[name] = valu
+
+def getLogExtra(**kwargs):
+    '''
+    Construct a properly enveloped log extra dictionary.
+
+    NOTE: If the key "exc" is specified, it will be used as
+          an exception to generate standardized error info.
+    '''
+    exc = kwargs.pop('exc', None)
+    extra = {'params': kwargs, 'loginfo': {}}
+
+    if exc is not None:
+        extra['loginfo']['error'] = s_common.excinfo(exc)
```

Comment:
`s_common.err()` output is more robust than `s_common.excinfo()`; repacking that output like how was done in `structlog.py` gave us useful traceback information. Was there a reason for using the older `excinfo` here ?

Reply:
Neither felt like exactly what i wanted, but `excinfo()` seemed a bit closer since it didn't require us to translate the first element of the tuple into a new dictionary key. I'd be totally fine with a completely log specific exception serialization format if we wanted to take the breaking-change opportunity to build out exactly what we want :+1: